From: =?utf-8?b?RmxvcmlhbiAic3Axcml0IuKAiw==?= <sp1rit@disroot.org>
Date: Fri, 11 Apr 2025 10:25:34 +0200
Subject: android: Add toplevel surface in bindNative

Previously, we were calling bindNative and attach the toplevel surface
after that suceeded in onCreate.
As we however also may end up calling (via proxy) bindNative in
AndroidToplevel.present - in case the previously registered activy
hasn't been bound yet - the Surface wasn't getting bound to the native
surface.

Part-of: <https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/8461>
(cherry picked from commit ee56ad314913137d965482bb6b4ca980c18e990c)

Origin: upstream gtk-4-18 after 4.18.4
---
 gdk/android/gdkandroidinit-private.h               |  1 +
 gdk/android/gdkandroidinit.c                       |  1 +
 gdk/android/gdkandroidtoplevel.c                   |  2 ++
 .../java/org/gtk/android/ToplevelActivity.java     | 28 ++++++++++++++--------
 4 files changed, 22 insertions(+), 10 deletions(-)

diff --git a/gdk/android/gdkandroidinit-private.h b/gdk/android/gdkandroidinit-private.h
index a731601..a8f7c46 100644
--- a/gdk/android/gdkandroidinit-private.h
+++ b/gdk/android/gdkandroidinit-private.h
@@ -77,6 +77,7 @@ typedef struct
     jfieldID native_identifier;
     jfieldID toplevel_view;
     jmethodID bind_native;
+    jmethodID attach_toplevel_surface;
     jmethodID post_window_configuration;
     jmethodID post_title;
   } toplevel;
diff --git a/gdk/android/gdkandroidinit.c b/gdk/android/gdkandroidinit.c
index 4255de4..8368918 100644
--- a/gdk/android/gdkandroidinit.c
+++ b/gdk/android/gdkandroidinit.c
@@ -230,6 +230,7 @@ gdk_android_initialize (JNIEnv *env, jobject application_classloader, jobject ac
   gdk_android_java_cache.toplevel.toplevel_view = (*env)->GetFieldID (env, gdk_android_java_cache.toplevel.klass, "view", "Lorg/gtk/android/ToplevelActivity$ToplevelView;");
   POPULATE_GDK_REFCACHE_STRING (toplevel, toplevel_identifier_key, "toplevelIdentifierKey")
   POPULATE_GDK_REFCACHE_METHOD (toplevel, bind_native, "bindNative", "(J)V")
+  POPULATE_GDK_REFCACHE_METHOD (toplevel, attach_toplevel_surface, "attachToplevelSurface", "()V")
   POPULATE_GDK_REFCACHE_METHOD (toplevel, post_window_configuration, "postWindowConfiguration", "(IZ)V")
   POPULATE_GDK_REFCACHE_METHOD (toplevel, post_title, "postTitle", "(Ljava/lang/String;)V")
   (*env)->RegisterNatives (env, toplevel_class, toplevel_natives, sizeof toplevel_natives / sizeof (JNINativeMethod));
diff --git a/gdk/android/gdkandroidtoplevel.c b/gdk/android/gdkandroidtoplevel.c
index c0e662c..fccf8c9 100644
--- a/gdk/android/gdkandroidtoplevel.c
+++ b/gdk/android/gdkandroidtoplevel.c
@@ -87,6 +87,8 @@ _gdk_android_toplevel_bind_native (JNIEnv *env, jobject this,
   self->activity = (*env)->NewGlobalRef (env, this);
   (*env)->SetLongField (env, this, gdk_android_get_java_cache ()->toplevel.native_identifier, native_identifier);
 
+  (*env)->CallVoidMethod (env, this, gdk_android_get_java_cache ()->toplevel.attach_toplevel_surface);
+
   gdk_android_toplevel_update_title (self);
   gdk_android_toplevel_update_window (self);
 
diff --git a/gdk/android/glue/java/org/gtk/android/ToplevelActivity.java b/gdk/android/glue/java/org/gtk/android/ToplevelActivity.java
index 25f30a3..7b2a8ab 100644
--- a/gdk/android/glue/java/org/gtk/android/ToplevelActivity.java
+++ b/gdk/android/glue/java/org/gtk/android/ToplevelActivity.java
@@ -44,6 +44,7 @@ import android.view.Window;
 import android.view.WindowInsets;
 import android.view.WindowInsetsController;
 
+import androidx.annotation.Keep;
 import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.annotation.UiThread;
@@ -177,7 +178,7 @@ public class ToplevelActivity extends Activity {
 			}
 
 			public void drop() {
-				if (this != ToplevelView.this.toplevel)
+				//if (this != ToplevelView.this.toplevel)
 					runOnUiThread(() -> ToplevelView.this.removeView(this));
 				/*else
 					logger.log(Level.SEVERE, "Attempted to drop toplevel from view");*/
@@ -418,19 +419,26 @@ public class ToplevelActivity extends Activity {
 				}
 			}
 
-			this.view.toplevel = this.view.new Surface(this.nativeIdentifier);
-			runOnUiThread(() -> {
-				this.view.addView(this.view.toplevel, this.view.new LayoutParams(
-						0, 0,
-						ViewGroup.LayoutParams.MATCH_PARENT,
-						ViewGroup.LayoutParams.MATCH_PARENT
-				));
-				this.view.requestLayout();
-			});
 			return null;
 		});
 	}
 
+	@Keep
+	@GlibContext.GtkThread
+	private void attachToplevelSurface() {
+		if (this.view.toplevel != null)
+			this.view.toplevel.drop();
+		this.view.toplevel = this.view.new Surface(this.nativeIdentifier);
+		runOnUiThread(() -> {
+			this.view.addView(this.view.toplevel, this.view.new LayoutParams(
+					0, 0,
+					ViewGroup.LayoutParams.MATCH_PARENT,
+					ViewGroup.LayoutParams.MATCH_PARENT
+			));
+			this.view.requestLayout();
+		});
+	}
+
 	private void updateToplevelState() {
 		boolean has_focus = hasWindowFocus();
 		boolean is_fullscreen = this.fullscreenState;
